docs/concepts.md

```diff
@@ -67,6 +68,96 @@ that are specific to a particular Node (like per-process UNIX file descriptors),
 so handle value 3 for Node A identify a different channel than handle value 3
 for Node B (unless there happens to be a coincidence of numbering).
 
+## Labels
+
+Labels are used to enforce information flow control (IFC) between nodes and
+channels in an Oak Application and the outside world.
+
+Each node and each channel is associated with a label at creation time. The
+label does not change any more after that (apart from explicit declassification
+operations, which are not currently supported).
+
+A label has two components: **secrecy** and **integrity**.
+
+Each component is represented as an (unordered) set of tags, and each tag refers
+to individual security principal in the system.
+
+The following types of security principals are supported in Oak:
+
+- **user**: specified by a bearer token, public key, or some other assertion of
+  a user identity
+- **computation**: specified by a measurement (hash) of the WebAssembly module
+  running within an Oak Node
+
+The least privileged label is "public_trusted" (also referred to as `⊥`,
+pronounced "bottom"), which corresponds to a node or channel which has only
+observed public data, and its inputs are not endorsed with any level of
+integrity; in this label, both secrecy and integrity components are empty sets.
+
+Given two labels `L_a` and `L_b`, `L_a ⊑ L_b` (pronounced "flows to") iff `L_a`
```

Comment:
Perhaps start a bit gentler with an intuitive description before writing the math. Also, although your description is technically correct (the best kind of correct), you might freak people out if you say the "... and less integrity" part because this does not capture the intuition as to why the order is this way. I think it is also valuable to explicitly give the definitions of the order on components which is not done presently. (Unless your intention is to make the fact that you are using the particular lattice order from set theory a hidden implementation detail, which case you might make other changes to further hide this).

Let's see if you like the following:

> Labels are compared to decide whether or not data movement is allowed, for example when a node sends a value over a channel. Given two labels `L_a` and `L_b`, `L_a ⊑ L_b` (pronounced "flows to") if a value labeled `L_a` can be sent to a destination labeled `L_b`. The "flows to" operator compares both confidentiality and integrity. The confidentiality check ensures that the destination is permitted to view any secrets that may have influenced the value. The integrity check ensures that the value sent is trusted by the destination.

> More concretely, if `L_a = (S_a, I_a)` (where `S_a` and `I_a` are the secrecy and integrity components respectively), and `L_b = (S_b, I_b)` then `L_a ⊑ L_b` iff `S_a ⊆ S_b` and `I_a ⊇I_b`. Notably, the integrity part of the check is "flipped" by convention [Biba1975]. More fundamental than adhering to convention, by flipping the integrity order in this way, we can retain the intuitive meaning behind integrity tags as representing trust by a principal (trust by a user, trust on a subject, etc.)

Reply:
Wow this looks great! A few minor questions:

- I noticed you use "secrecy" and "confidentiality" interchangeably; Flume and Flam seem to use them respectively; is one of them preferred in current literature, or anyways do you think one would be more accurate than the other? Let me know which one you prefer (I think we should pick one, for consistency), and I'm happy to make changes elsewhere in this doc and in the code, one way or the other.

- What's "Biba1975"? I assume it's a citation reference, but I didn't manage to resolve it. Do you have an actual URL I could use?

Reply:
Confidentiality and secrecy appear to be used interchangeably in the literature as far as I know. You can pick whichever you like and I'll try to be consistent with you. It looks like maybe secrecy?

The paper is this one: http://seclab.cs.ucdavis.edu/projects/history/papers/biba75.pdf which contains the Biba integrity model. This is highly-cited early work that is generally attributed as the first to point out that integrity is the dual of confidentiality. It might not actually be that useful to really read it though (although I am a person that enjoys reading very old papers).